README.md
# NAME

Pod::Elemental::Transformer::Stenciller - Injects content from textfiles transformed with Stenciller

<div>
    <p>
    <img src="https://img.shields.io/badge/perl-5.14+-blue.svg" alt="Requires Perl 5.14+" />
    <img src="https://img.shields.io/badge/coverage-88.6%25-orange.svg" alt="coverage 88.6%" />
    <a href="https://github.com/Csson/p5-Pod-Elemental-Transformer-Stenciller/actions?query=workflow%3Amakefile-test"><img src="https://img.shields.io/github/workflow/status/Csson/p5-Pod-Elemental-Transformer-Stenciller/makefile-test" alt="Build status at Github" /></a>
    </p>
</div>

# VERSION

Version 0.0301, released 2021-06-30.

# SYNOPSIS

    # in weaver.ini
    [-Transformer / Stenciller]
    transformer = Stenciller
    directory = path/to/stencildir

# DESCRIPTION

This transformer uses a special command in pod files to inject content from elsewhere via a [Stenciller](https://metacpan.org/pod/Stenciller) transformer plugin.

## Example

1\. Start with the `weaver.ini` from the ["synopsis"](#synopsis).

2\. Add a textfile, in `path/to/stencildir/file-with-stencils.stencil`:

    == stencil { to_pod => 1 } ==

    Header text

    --input--

        Input text

    --end input--

    Between text

    --output--

        Output text

    --end output--

    Footer text

3\. Add a Perl module:

    package A::Test::Module;

    1;

    __END__

    =pod

    =head1 NAME

    =head1 DESCRIPTION

    :stenciller ToUnparsedText file-with-stencils.stencil

The last line in the Perl module will result in the following:

- The textfile is parsed with [Stenciller](https://metacpan.org/pod/Stenciller)
- The textfile is then transformed using the [Stenciller::Plugin::ToUnparsedText](https://metacpan.org/pod/Stenciller::Plugin::ToUnparsedText) plugin.
- The ':stenciller ...' line in the pod is replaced with the transformed content.

## Pod hash

It is possible to filter stencils by index with an optional hash in the pod:

    :stenciller ToUnparsedText 1-test.stencil { stencils => [0, 2..4] }

This will only include the stencils with index 0, 2, 3 and 4 from `file-with-stencils.stencil`.

## Stencil hash

This module checks for the `to_pod` key in the stencil hash. If it has a true value (or doesn't exist) it is included in the transformation.

However, any stencil excluded by the ["Pod hash"](#pod-hash) is already disregarded. It is probably least confusing to choose **one** of these places to do all filtering.

# ATTRIBUTES

## directory

<table cellpadding="0" cellspacing="0">
<tr>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;"><a href="https://metacpan.org/pod/Types::Path::Tiny#Dir">Dir</a></td>
    <td style="padding-right: 6px; padding-left: 6px; border-right: 1px solid #b8b8b8; white-space: nowrap;">required</td>
    <td style="padding-left: 6px; padding-right: 6px; white-space: nowrap;">read-only</td>
</tr>
</table>

<p>Path to directory where the stencil files are.</p>

# SEE ALSO

- [Stenciller](https://metacpan.org/pod/Stenciller)
- [Stenciller::Plugin::ToUnparsedText](https://metacpan.org/pod/Stenciller::Plugin::ToUnparsedText)
- [Pod::Weaver](https://metacpan.org/pod/Pod::Weaver)

# SOURCE

[https://github.com/Csson/p5-Pod-Elemental-Transformer-Stenciller](https://github.com/Csson/p5-Pod-Elemental-Transformer-Stenciller)

# HOMEPAGE

[https://metacpan.org/release/Pod-Elemental-Transformer-Stenciller](https://metacpan.org/release/Pod-Elemental-Transformer-Stenciller)

# AUTHOR

Erik Carlsson <info@code301.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2021 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Pod-Elemental-Transformer-Stenciller

{{$NEXT}}
 [API Changes]

 [Bug Fixes]

 [Documentation]

 [Enhancements]

 [New Features]

 [Other Changes]

 [Requirements]

 [Stats & Dependency Changes]

 [Testing]

0.0301 2021-06-30T18:24:25Z UTC
 [Other Changes]
 - Explicit $VERSION
 - Replace Travis CI with Github Actions

 [Stats & Dependency Changes]
 - 3 files changed, 63 insertions(+), 1 deletion(-)
 - (run req) ~ Carp (any) --> 1.38
 - (run req) ~ Module::Load (any) --> 0.22
 - (run req) ~ Path::Tiny (any) --> 0.072
 - (run req) ~ Types::Path::Tiny (any) --> 0.005
 - (dev req) + Badge::Depot 0.0103
 - (dev req) + Badge::Depot::Plugin::Coverage (any)
 - (dev req) + Badge::Depot::Plugin::Githubactions (any)
 - (dev req) + Badge::Depot::Plugin::Perl (any)
 - (dev req) + Dist::Zilla 6.015
 - (dev req) +
   Dist::Zilla::Plugin::Author::CSSON::GithubActions::Workflow::TestWithMakefile
   0.0106
 - (dev req) + Dist::Zilla::Plugin::Authority 1.009
 - (dev req) + Dist::Zilla::Plugin::BumpVersionAfterRelease::Transitional
   0.009
 - (dev req) + Dist::Zilla::Plugin::ChangeStats::Dependencies::Git 0.0200
 - (dev req) + Dist::Zilla::Plugin::Git 2.046
 - (dev req) + Dist::Zilla::Plugin::Git::CheckFor::CorrectBranch 0.014
 - (dev req) + Dist::Zilla::Plugin::Git::Commit (any)
 - (dev req) + Dist::Zilla::Plugin::MetaConfig (any)
 - (dev req) + Dist::Zilla::Plugin::NextRelease::Grouped 0.0200
 - (dev req) + Dist::Zilla::Plugin::PromptIfStale 0.057
 - (dev req) + Dist::Zilla::Plugin::RewriteVersion::Transitional 0.007
 - (dev req) + Pod::Weaver::Section::GenerateSection 1.01
 - (dev req) + Test::EOF (any)
 - (dev req) + Test::Version 1
 - (dev req) ~ Dist::Zilla::Plugin::CheckChangesHasContent (any) --> 0.011
 - (dev req) ~ Dist::Zilla::Plugin::Clean (any) --> 0.07
 - (dev req) ~ Dist::Zilla::Plugin::CopyFilesFromBuild (any) --> 0.170880
 - (dev req) ~ Dist::Zilla::Plugin::Git::Contributors (any) --> 0.035
 - (dev req) ~ Dist::Zilla::Plugin::GithubMeta (any) --> 0.54
 - (dev req) ~ Dist::Zilla::Plugin::InstallRelease (any) --> 0.008
 - (dev req) ~ Dist::Zilla::Plugin::MetaProvides::Package (any) -->
   2.004003
 - (dev req) ~ Dist::Zilla::Plugin::PodWeaver (any) --> 4.006
 - (dev req) ~ Dist::Zilla::Plugin::ReadmeAnyFromPod (any) --> 0.163250
 - (dev req) ~ Dist::Zilla::Plugin::RunExtraTests (any) --> 0.029
 - (dev req) ~ Dist::Zilla::Plugin::Test::Compile (any) --> 2.058
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOF (any) --> 0.0501
 - (dev req) ~ Dist::Zilla::Plugin::Test::EOL (any) --> 0.18
 - (dev req) ~ Dist::Zilla::Plugin::Test::NoTabs (any) --> 0.15
 - (dev req) ~ Dist::Zilla::Plugin::Test::ReportPrereqs (any) --> 0.027
 - (dev req) ~ Dist::Zilla::Plugin::Test::Version (any) --> 1.09
 - (dev req) ~ Pod::Elemental::Transformer::Splint (any) --> 0.1201
 - (dev req) ~ Pod::Weaver::Section::Badges (any) --> 0.0401
 - (dev req) ~ Test::NoTabs (any) --> 1.4
 - (dev req) ~ Test::Warnings (any) --> 0.026
 - (dev req) - Dist::Iller::Config::Author::CSSON
 - (dev req) - Dist::Zilla::Plugin::MetaProvides::Class
 - (dev req) - Dist::Zilla::Plugin::NextRelease
 - (dev req) - Dist::Zilla::Plugin::NextVersion::Semantic
 - (dev req) - Dist::Zilla::Plugin::OurPkgVersion
 - (dev req) - Dist::Zilla::Plugin::PodnameFromClassname
 - (dev req) - Dist::Zilla::Plugin::Prereqs::Plugins
 - (dev req) - Dist::Zilla::Plugin::PreviousVersion::Changelog
 - (dev req) - Dist::Zilla::Plugin::ReversionOnRelease
 - (dev req) - Dist::Zilla::Plugin::Test::Kwalitee::Extra
 - (dev req) - Dist::Zilla::Plugin::TravisYML
 - (dev req) - Pod::Weaver::Section::Homepage::DefaultCPAN
 - (dev req) - Pod::Weaver::Section::Source::DefaultGitHub
 - (dev sug) + Dist::Iller 0.1411
 - (dev sug) + Dist::Iller::Config::Author::CSSON 0.0328

0.0300 2016-02-02T15:05:04Z UTC
 [Requirements]
 - Removed Kavorka/Moops dependency, so will work on 5.10+ (once Stenciller
   does).
 - No longer uses Stenciller::Standard for dependencies (but uses them
   directly).

0.0200 2015-11-23T00:10:51Z UTC
 [Documentation]
 - Minor documentation improvements

 [Meta]
 - Added .travis.yml.

 [Requirements]
 - Bump Stenciller requirement to 0.1300.

0.0101 2015-02-13T01:04:10Z CET
 [Documentation]
 - Added documentation regarding the option hashes in pod and stencil.
 - Corrected package name in Changes

 [Enhancements]
 - Added check for 'to_pod' key in stencil hash.

 [Meta]
 - Removed cpanfile (now prereqs are set in dist.ini).

 [Requirements]
 - Bump Stenciller prereq to 0.1212.

0.0100 2015-02-07T15:45:56Z CET
 - Initial release


t/corpus/lib/Test/For/StencillerFromUnparsedText.pm
package Test::For::StencillerFromUnparsedText;

1;

__END__

=pod

=head1 NAME

=head1 DESCRIPTION

:stenciller ToUnparsedText 1-test.stencil { stencils => [0, 2..4] }


lib/Pod/Elemental/Transformer/Stenciller.pm
use 5.10.1;
use strict;
use warnings;

package Pod::Elemental::Transformer::Stenciller;

# ABSTRACT: Injects content from textfiles transformed with Stenciller
# AUTHORITY
our $VERSION = '0.0302';

use Moose;
use MooseX::AttributeDocumented;
with qw/Pod::Elemental::Transformer Stenciller::Utils/;

use namespace::autoclean;
use Types::Standard -types;
use Types::Path::Tiny qw/Dir/;
use Types::Stenciller qw/Stenciller/;
use Carp qw/carp croak/;
use Module::Load qw/load/;
use Path::Tiny;
use Stenciller;

has directory => (
    is => 'ro',
    isa => Dir,
    coerce => 1,
    required => 1,
    documentation => 'Path to directory where the stencil files are.'
);
has settings => (
    is => 'ro',
    isa => HashRef,
    traits => ['Hash'],
    default => sub { +{} },
    handles => {
        get_setting => 'get',
        set_setting => 'set',
        all_settings => 'elements',
    },
    documentation_default => '{ }',
    documentation_order => 0,
    documentation => 'If a plugin takes more attributes..',
);
has plugins => (
    is => 'ro',
    isa => HashRef,
    default => sub { +{} },
    documentation_default => '{ }',
    documentation_order => 0,
    init_arg => undef,
    traits => ['Hash'],
    handles => {
        get_plugin => 'get',
        set_plugin => 'set',
    },
    documentation_order => 0,
);
has stencillers => (
    is => 'ro',
    isa => HashRef,
    traits => ['Hash'],
    init_arg => undef,
    handles => {
        get_stenciller_for_filename => 'get',
        set_stenciller_for_filename => 'set',
    },
    documentation_order => 0,
);

around BUILDARGS => sub {
    my $next = shift;
    my $class = shift;
    my @args = @_;

    my $args = ref $args[0] eq 'HASH' ? $args[0] : { @args };

    $class->$next(
        directory => delete $args->{'directory'},
        settings  => $args
    );
};

sub transform_node {
    my $self = shift;
    my $main_node = shift;

    NODE:
    foreach my $node (@{ $main_node->children }) {
        my $content = $node->content;
        my $start = substr($content, 0, 11, '');
        next NODE if $start ne ':stenciller';

        $content =~ s{^\h+}{};             # remove leading whitespace
        next if $content !~ m{([^\h\v]+)}; # the next sequence of non-space is the wanted plugin name

        my $wanted_plugin = $1;
        my $plugin_name = $self->ensure_plugin($wanted_plugin);

        (undef, my($filename, $possible_hash)) = split /\h+/ => $content, 3;
        chomp $filename;
        my $node_settings = defined $possible_hash && $possible_hash =~ m{\{.*\}} ? $self->eval_to_hashref($possible_hash, $filename) : {};

        my $stenciller = $self->get_stenciller_for_filename($filename);

        if(!Stenciller->check($stenciller)) {
            $stenciller = Stenciller::->new(filepath => path($self->directory)->child($filename));
            if(!$stenciller->has_stencils) {
                carp sprintf '! no stencils in %s/%s - skipping', $self->directory, $filename;
                return;
            }

            $self->set_stenciller_for_filename($filename => $stenciller);
        }

        my $transformed_content = $stenciller->transform(plugin_name => $plugin_name,
                                                         constructor_args => $self->settings,
                                                         transform_args => { %$node_settings, require_in_extra => { key => 'to_pod', value => 1, default => 1 } },
                                                        );
        $transformed_content =~ s{[\v\h]+$}{};
        $node->content($transformed_content);

    }
}
sub ensure_plugin {
    my $self = shift;
    my $plugin_name = shift;

    return $self->get_plugin($plugin_name) if $self->get_plugin($plugin_name);

    my $plugin_class = sprintf 'Stenciller::Plugin::%s', $plugin_name;
    load($plugin_class);

    if(!$plugin_class->does('Stenciller::Transformer')) {
        croak("[$plugin_name] doesn't do the Stenciller::Transformer role. Quitting.");
    }
    $self->set_plugin($plugin_name => $plugin_name);
    return $plugin_name;
}

__PACKAGE__->meta->make_immutable;

1;

=pod

:splint classname Pod::Elemental::Transformer::Stenciller

=head1 SYNOPSIS

    # in weaver.ini
    [-Transformer / Stenciller]
    transformer = Stenciller
    directory = path/to/stencildir

=head1 DESCRIPTION

This transformer uses a special command in pod files to inject content from elsewhere via a L<Stenciller> transformer plugin.

=head2 Example

1. Start with the C<weaver.ini> from the L</"synopsis">.

2. Add a textfile, in C<path/to/stencildir/file-with-stencils.stencil>:

    == stencil { to_pod => 1 } ==

    Header text

    --input--

        Input text

    --end input--

    Between text

    --output--

        Output text

    --end output--

    Footer text

3. Add a Perl module:

    package A::Test::Module;

    1;

    __END__

    =pod

    =head1 NAME

    =head1 DESCRIPTION

    :stenciller ToUnparsedText file-with-stencils.stencil

The last line in the Perl module will result in the following:

=for :list
* The textfile is parsed with L<Stenciller>
* The textfile is then transformed using the L<Stenciller::Plugin::ToUnparsedText> plugin.
* The ':stenciller ...' line in the pod is replaced with the transformed content.

=head2 Pod hash

It is possible to filter stencils by index with an optional hash in the pod:

    :stenciller ToUnparsedText 1-test.stencil { stencils => [0, 2..4] }

This will only include the stencils with index 0, 2, 3 and 4 from C<file-with-stencils.stencil>.

=head2 Stencil hash

This module checks for the C<to_pod> key in the stencil hash. If it has a true value (or doesn't exist) it is included in the transformation.

However, any stencil excluded by the L</"Pod hash"> is already disregarded. It is probably least confusing to choose B<one> of these places to do all filtering.

=head1 ATTRIBUTES

:splint attributes

=head1 SEE ALSO

=for :list
* L<Stenciller>
* L<Stenciller::Plugin::ToUnparsedText>
* L<Pod::Weaver>

=cut


lib/Pod/Weaver/Transformer/Stenciller/FromUnparsedText.pm
use Stenciller::Standard;

# VERSION
# PODCLASSNAME
# ABSTRACT: Injects text parsed with Stenciller::Plugin::ToUnparsed::Text

class Pod::Elemental::Transformer::Stenciller::FromUnparsedText
using Moose
 with Pod::Elemental::Transformer {

    has directory => (
        is => 'ro',
        isa => File,
        required => 1,
    );
    has filepattern => (
        is => 'rw',
        isa => RegexpRef,
        default => sub { qr/^\w-\d+\.stencil$/ },
    );
    
    

    method transform_node {

    }

}







1;


__END__

=pod

:splint classname Pod::Elemental::Transformer::Stenciller::FromUnparsedText

=head1 SYNOPSIS

    use Pod::Weaver::Plugin::Stenciller::FromUnparsedText;

=head1 DESCRIPTION

Pod::Weaver::Plugin::Stenciller::FromUnparsedText is ...

=head1 ATTRIBUTES

:splint attributes

=head1 SEE ALSO

=cut


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.014000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Injects content from textfiles transformed with Stenciller",
  "AUTHOR" => "Erik Carlsson <info\@code301.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Pod-Elemental-Transformer-Stenciller",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.014000",
  "NAME" => "Pod::Elemental::Transformer::Stenciller",
  "PREREQ_PM" => {
    "Carp" => "1.38",
    "Module::Load" => "0.22",
    "Moose" => "2.1400",
    "MooseX::AttributeDocumented" => "0.1003",
    "Path::Tiny" => "0.072",
    "Pod::Elemental::Transformer" => 0,
    "Stenciller" => "0.1302",
    "Stenciller::Utils" => 0,
    "Types::Path::Tiny" => "0.005",
    "Types::Standard" => "1.000000",
    "Types::Stenciller" => 0,
    "namespace::autoclean" => "0.24"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Pod::Elemental" => 0,
    "Pod::Elemental::Transformer::Pod5" => 0,
    "Test::Differences" => 0,
    "Test::More" => "0.96"
  },
  "VERSION" => "0.0302",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => "1.38",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Load" => "0.22",
  "Moose" => "2.1400",
  "MooseX::AttributeDocumented" => "0.1003",
  "Path::Tiny" => "0.072",
  "Pod::Elemental" => 0,
  "Pod::Elemental::Transformer" => 0,
  "Pod::Elemental::Transformer::Pod5" => 0,
  "Stenciller" => "0.1302",
  "Stenciller::Utils" => 0,
  "Test::Differences" => 0,
  "Test::More" => "0.96",
  "Types::Path::Tiny" => "0.005",
  "Types::Standard" => "1.000000",
  "Types::Stenciller" => 0,
  "namespace::autoclean" => "0.24"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
